File: packages/velox_connectivity/android/src/main/kotlin/com/velox/connectivity/VeloxConnectivityPlugin.kt
package com.velox.connectivity

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class VeloxConnectivityPlugin : FlutterPlugin, MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var connectivityManager: ConnectivityManager? = null
    private var networkCallback: ConnectivityManager.NetworkCallback? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        connectivityManager =
            context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        methodChannel = MethodChannel(binding.binaryMessenger, "com.velox.connectivity/method")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "com.velox.connectivity/event")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        unregisterNetworkCallback()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "checkConnectivity" -> result.success(getConnectivityInfo())
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        registerNetworkCallback()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
        unregisterNetworkCallback()
    }

    private fun getConnectivityInfo(): Map<String, Any> {
        val cm = connectivityManager ?: return mapOf(
            "status" to "unknown",
            "type" to "unknown",
            "isOnline" to false,
        )

        val network = cm.activeNetwork
        val capabilities = if (network != null) cm.getNetworkCapabilities(network) else null

        if (capabilities == null) {
            return mapOf(
                "status" to "disconnected",
                "type" to "none",
                "isOnline" to false,
            )
        }

        val isOnline = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)

        val type = when {
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "mobile"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH) -> "bluetooth"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> "vpn"
            else -> "unknown"
        }

        return mapOf(
            "status" to if (isOnline) "connected" else "disconnected",
            "type" to type,
            "isOnline" to isOnline,
        )
    }

    private fun registerNetworkCallback() {
        val cm = connectivityManager ?: return
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                eventSink?.success(getConnectivityInfo())
            }

            override fun onLost(network: Network) {
                eventSink?.success(
                    mapOf(
                        "status" to "disconnected",
                        "type" to "none",
                        "isOnline" to false,
                    )
                )
            }

            override fun onCapabilitiesChanged(
                network: Network,
                capabilities: NetworkCapabilities
            ) {
                eventSink?.success(getConnectivityInfo())
            }
        }
        networkCallback = callback
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        cm.registerNetworkCallback(request, callback)
    }

    private fun unregisterNetworkCallback() {
        networkCallback?.let { connectivityManager?.unregisterNetworkCallback(it) }
        networkCallback = null
    }
}

File: packages/velox_connectivity/ios/Classes/VeloxConnectivityPlugin.swift
import Flutter
import Network

public class VeloxConnectivityPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var pathMonitor: NWPathMonitor?
    private let monitorQueue = DispatchQueue(label: "com.velox.connectivity.monitor")

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VeloxConnectivityPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "com.velox.connectivity/method",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "com.velox.connectivity/event",
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "checkConnectivity":
            checkConnectivity(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        startMonitoring()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        stopMonitoring()
        self.eventSink = nil
        return nil
    }

    private func checkConnectivity(result: @escaping FlutterResult) {
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            monitor.cancel()
            let info = self?.pathToMap(path) ?? [
                "status": "unknown",
                "type": "unknown",
                "isOnline": false,
            ]
            DispatchQueue.main.async {
                result(info)
            }
        }
        monitor.start(queue: monitorQueue)
    }

    private func startMonitoring() {
        stopMonitoring()
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            let info = self?.pathToMap(path) ?? [:]
            DispatchQueue.main.async {
                self?.eventSink?(info)
            }
        }
        monitor.start(queue: monitorQueue)
        pathMonitor = monitor
    }

    private func stopMonitoring() {
        pathMonitor?.cancel()
        pathMonitor = nil
    }

    private func pathToMap(_ path: NWPath) -> [String: Any] {
        let isOnline = path.status == .satisfied
        let type: String
        if path.usesInterfaceType(.wifi) {
            type = "wifi"
        } else if path.usesInterfaceType(.cellular) {
            type = "mobile"
        } else if path.usesInterfaceType(.wiredEthernet) {
            type = "ethernet"
        } else if path.usesInterfaceType(.loopback) {
            type = "unknown"
        } else if path.usesInterfaceType(.other) {
            type = "unknown"
        } else {
            type = "none"
        }

        return [
            "status": isOnline ? "connected" : "disconnected",
            "type": type,
            "isOnline": isOnline,
        ]
    }
}
